src/libcrun/criu.c

```diff
@@ -218,6 +218,88 @@ libcrun_container_checkpoint_linux_criu (libcrun_container_status_t *status,
   return 0;
 }
 
+static int
+prepare_restore_mounts (runtime_spec_schema_config_schema *def,
+                        char *root,
+                        libcrun_error_t *err)
+{
+  int i;
+
+  /* Go through all mountpoints to be able to recreate missing mountpoints. */
+  for (i = 0; i < def->mounts_len; i++)
+    {
+      char *dest = def->mounts[i]->destination;
+      char *type = def->mounts[i]->type;
+      cleanup_close int root_fd = -1;
+      bool on_tmpfs = false;
+      int is_dir = 1;
+      size_t j;
+
+      /* cgroup restore should be handled by CRIU itself */
```

Comment:
That is oversimplification :)

Also I believe in case of cgroup v2 the /sys/fs/cgroup needs to be created (see e.g. https://github.com/opencontainers/runc/pull/2335/commits/f7d0e37a5881de18cae9f252e5014f1ba1c5f4bd#diff-7b8effb45402944e445a664e4d9c296dR1188)

have you tested this on F31 with cgroup v2 unified (i.e. *without* `systemd.unified_cgroup_hierarchy=0`)? From a quick glance it should not work unless /sys/fs/cgroup mp is created.

Reply:
I actually have only tested on F31 with cgroup2. Mounting `/sys` always provides `/sys/fs/cgroup`, so the comment is wrong, but it should work for cgroup1 and cgroup2. Actually, let me test i quickly... Yes, rebooted F31 with cgroup1 and it also works.

Works for cgroup1 as `/sys/fs/cgroup` is created by mounting sysfs and all other cgroup mountpoints are in a tmpfs restored by CRIU. For cgroup2 it is only one mountpoint: `/sys/fs/cgroup`

No idea if we need to think about the case of no sysfs but a mounted cgroup. That would not work.

Reply:
Yes, I was dead wrong about the need to create a mountpoint for `/sys/fs/cgroup`.